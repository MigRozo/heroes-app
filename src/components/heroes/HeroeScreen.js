import React, { useMemo } from 'react';
import { Redirect, useParams } from 'react-router-dom';
import { getHeroesById } from '../../selectors/getHeroeById';

// import batman from '../../assets/heroes/dc-batman.jpg'; //recurso estÃ¡tico
// const heroeImages = require.context('../../assets/heroes', true);
import { heroeImages } from '../../helpers/heroeImages';

export const HeroeScreen = ({ history }) => {
    const { heroeId } = useParams();

    // const heroe = getHeroesById( heroeId );
    const heroe = useMemo(() => getHeroesById( heroeId ), [ heroeId ]);
    if ( !heroe ) return <Redirect to="/" />;

    const handleReturn = () => {
        if ( history.length <= 2 ) {
            history.push('/');
        } else {
            history.goBack();
        }
    };

    const { superhero, publisher, alter_ego, first_appearance, characters } = heroe;

    return (
        <div className="row mt-5">
            <div className="col-4">
                <img
                    // src={ `../assets/heroes/${ heroeId }.jpg` } //desde public/assets
                    // src={ batman } // con import
                    src={ heroeImages(`./${ heroeId }.jpg`).default }
                    alt={ superhero }
                    className="img-thumbnail  animate__animated animate__fadeInLeft"
                />
            </div>
            <div className="col-8 animate__animated animate__fadeIn">
                <h3>{ superhero }</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"><strong>Alter ego:</strong> { alter_ego }</li>
                    <li className="list-group-item"><strong>Publisher:</strong> { publisher }</li>
                    <li className="list-group-item"><strong>First appearance:</strong> { first_appearance }</li>
                </ul>
                <h5>Characters</h5>
                <p>{ characters }</p>

                <button onClick={ handleReturn } className="btn btn-primary">Return</button>
            </div>
        </div>
    );
};
