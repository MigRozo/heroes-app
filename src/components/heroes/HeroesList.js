import React, { useMemo } from 'react';
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher';
import { HeroesCard } from './HeroesCard';

export const HeroesList = ({ publisher }) => {

    // const heroes = getHeroesByPublisher( publisher );
    const heroes = useMemo(() => getHeroesByPublisher( publisher ), [ publisher ]);

    return (
        <div className="card-columns">
            {
                heroes.map( heroe => (
                    <HeroesCard
                        key={ heroe.id }
                        { ...heroe }
                    />
                ))
            }
        </div>
    );
};
