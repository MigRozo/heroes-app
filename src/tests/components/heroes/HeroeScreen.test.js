import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';

import { HeroeScreen } from '../../../components/heroes/HeroeScreen';

describe('Pruebas en <HeroeScreen />', () => {

    const historyMock = {
        length: 10,
        push: jest.fn(),
        goBack: jest.fn()
    };

    // heroe/dc-batman

    test('Debe mostrar el componente Redirect si no hay argumentos en URL', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe']}>
                <HeroeScreen history={ historyMock } />
            </MemoryRouter>
        );

        // expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('Redirect').exists() ).toBe( true );
    });

    test('Debe mostrar un heroe si el parametro existe y se encuentra', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/dc-batman']}>
                <Route path="/heroe/:heroeId" component={ HeroeScreen } />
            </MemoryRouter>
        );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('.row').exists() ).toBe( true );
    });

    test('Debe regresar a la pantalla anterior con push', () => {
        const historyMock = {
            length: 1,
            push: jest.fn(),
            goBack: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/dc-batman']}>
                <Route path="/heroe/:heroeId" component={ (props) => <HeroeScreen history={ historyMock } /> } />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();
        expect( historyMock.push ).toHaveBeenCalledWith('/');
        expect( historyMock.goBack ).not.toHaveBeenCalled();

    });

    test('Debe regresar a la pantalla anterior GOBACK', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/dc-batman']}>
                <Route path="/heroe/:heroeId" component={ (props) => <HeroeScreen history={ historyMock } /> } />
            </MemoryRouter>
        );

        wrapper.find('button').prop('onClick')();
        // expect( historyMock.push ).not.toHaveBeenCalled();
        expect( historyMock.push ).toHaveBeenCalledTimes(0);
        expect( historyMock.goBack ).toHaveBeenCalled();

    });

    test('Debe llamar el redirect si el heroe no existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/heroe/dc-spiderman616']}>
                <Route path="/heroe/:heroeId" component={ (props) => <HeroeScreen history={ historyMock } /> } />
            </MemoryRouter>
        );

        expect( wrapper.text().trim() ).toBe('');

    });

});
