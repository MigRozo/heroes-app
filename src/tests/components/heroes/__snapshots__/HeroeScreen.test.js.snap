// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroeScreen /> Debe mostrar un heroe si el parametro existe y se encuentra 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Batman"
      className="img-thumbnail  animate__animated animate__fadeInLeft"
      src="../assets/heroes/dc-batman.jpg"
    />
  </div>
  <div
    className="col-8 animate__animated animate__fadeIn"
  >
    <h3>
      Batman
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <strong>
          Alter ego:
        </strong>
         
        Bruce Wayne
      </li>
      <li
        className="list-group-item"
      >
        <strong>
          Publisher:
        </strong>
         
        DC Comics
      </li>
      <li
        className="list-group-item"
      >
        <strong>
          First appearance:
        </strong>
         
        Detective Comics #27
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Bruce Wayne
    </p>
    <button
      className="btn btn-primary"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
